service/dap/server.go

```diff
@@ -130,6 +130,18 @@ type Server struct {
 	// sendingMu synchronizes writing to net.Conn
 	// to ensure that messages do not get interleaved
 	sendingMu sync.Mutex
+
+	// running tracks the running state according to the server,
+	// which may not correspond to the actual running state of
+	// the process (e.g. if a running command is temporarily interrupted).
+	running   bool
+	runningMu sync.Mutex
+
+	// changeStateMu must be held for a request to resume execution of the
+	// process.
+	// We use this to guarantee that no other request can unexpectedly
```

Comment:
I haven't considered this. 

In this case I am trying to cover two concepts:
1) whether we already have an async command running
2) internally changing the running state of the process

It is entirely possible this could be covered by a RWMutex, but I had not been looking at it. I will think a bit about whether this could be a better mechanism.